### PR TITLE
Feature/defensive coach

### PR DESCRIPTION
**Pull request by rolfvdhulst**
- [x] The tests are passing

### Summary
Assigns defensive positions and defenders that go there. Still far from complete, but should be a good start for passive defenders, much like how the 'passive' keeper works half-decently. Especially work integrating the ball possession and pass detection is still required, though the first starts for that should be in other branches already. 
### Related issues
If there are related issues, reference them here.

fixes ...  
fixes ...

### Comments for the reviewer
If there are comments for the reviewer, leave them here

### Checklist for the reviewer
This acts as a guide to help you and/or the reviewer to analyze the code. 

**functionality**  
- Is the code is properly tested?
- Is the code working as expected
- Is the code is understandable and easy to read
- Are proper abstractions made? (no duplicate code, nice use of functions/objects)

**style**  
- Variables/objects/classes have proper names
- Proper use of variables: nu unused vars, no magic numbers
- No possible memory leaks. Mutexes properly used when multithreading? possible pointer issues?
- Are smart pointers used?
- unused variables?
- magic numbers?
- well documented?
- No unused imports?
- No leftover code / unused comments
- Proper indentation
- Proper use of namespaces

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -196,7 +198,7 @@ set(SKILLS_SOURCES
         ${PROJECT_SOURCE_DIR}/roboteam_ai/src/skills/EnterFormation.cpp
         ${PROJECT_SOURCE_DIR}/roboteam_ai/src/skills/EnterFormation.h
         ${PROJECT_SOURCE_DIR}/roboteam_ai/src/skills/AvoidBall.cpp
-        ${PROJECT_SOURCE_DIR}/roboteam_ai/src/skills/AvoidBall.h)
+        ${PROJECT_SOURCE_DIR}/roboteam_ai/src/skills/AvoidBall.h roboteam_ai/src/skills/CoachDefend.cpp roboteam_ai/src/skills/CoachDefend.h)
```
Comment (mrlukasbos) [inline CMakeLists.txt]: ` ${PROJECT_SOURCE_DIR}`

File: roboteam_ai/src/control/ControlUtils.cpp [hunk 2]
```diff
@@ -232,7 +239,18 @@ Vector2 ControlUtils::twoLineIntersection(Vector2 a1, Vector2 a2, Vector2 b1, Ve
         return Vector2();
 
 }
+/// returns true if the line intersects in the positive extension from point a1 to a2 with the extended line through b1 and b2
+double ControlUtils::twoLineForwardIntersection(Vector2 a1, Vector2 a2, Vector2 b1, Vector2 b2) {
```
Comment (mrlukasbos) [inline roboteam_ai/src/control/ControlUtils.cpp]: is this tested?

Comment (mrlukasbos) [reply roboteam_ai/src/control/ControlUtils.cpp]: if not, this definitely needs a test

File: roboteam_ai/src/control/ControlUtils.cpp [hunk 3]
```diff
@@ -286,6 +304,10 @@ std::vector<std::pair<Vector2, Vector2>> ControlUtils::calculateClosestPathsFrom
     return solutionPairs;
 }
 
+Vector2 ControlUtils::computeSimpleReceivePos(Vector2 startPos, Vector2 robotPos){
```
Comment (mrlukasbos) [inline roboteam_ai/src/control/ControlUtils.cpp]: this is not something that belongs in controlUtils I think. Most controlUtils are about vector calculus etc.

File: roboteam_ai/src/control/ControlUtils.cpp (hunk 2 above)
Comment (rolfvdhulst) [reply roboteam_ai/src/control/ControlUtils.cpp]: wrote a test now

File: roboteam_ai/src/control/ControlUtils.cpp (hunk 3 above)
Comment (rolfvdhulst) [reply roboteam_ai/src/control/ControlUtils.cpp]: moved it

File: roboteam_ai/src/coach/DefensiveCoach.cpp [hunk 4]
```diff
@@ -0,0 +1,471 @@
...
+    defenderLocations.clear();
+    std::vector<int> availableDefenders = defenders;
+    // decide the locations to defend
+    std::vector<std::pair<Vector2, double>> positions = decideDefendersOnDefenseLine(availableDefenders.size());
+    // the following algorithm takes the closest robot for each available defender to decide which robot goes where.
+    // Since the points are ordered on priority from the above algorithm the most important points come first
+    // It might be better to use an algorithm that is more complicated (e.g. hungarian) but then we might need some kind of system which gives the first points more 'priority'
+    for (auto position : positions) {
+        int bestId = - 1;
+        double bestDist = 10000000000;
```
Comment (RobZelluf) [inline roboteam_ai/src/coach/DefensiveCoach.cpp]: Use INT_MAX

File: roboteam_ai/src/coach/DefensiveCoach.cpp [hunk 5]
```diff
@@ -0,0 +1,471 @@
...
+            availableDefenders.erase(std::find(availableDefenders.begin(), availableDefenders.end(), bestId));
+        }
+        else {
+            std::cerr << "Could not find a robot to defend location!!!";
+            return;
+        }
+    }
+    //visualization
+    visualizePoints();
+    auto stop = std::chrono::high_resolution_clock::now();
```
Comment (RobZelluf) [inline roboteam_ai/src/coach/DefensiveCoach.cpp]: Is clocking this still needed?

**Comment by mrlukasbos** [comment]
replaced by #252

**Pull request closed without merging**